keps/sig-node/1664-node-ips/README.md

```diff
@@ -0,0 +1,737 @@
+     sorting associated with that.
+
+  2. If `--node-ips` has a single element, then all addresses of the
+     non-matching family will be removed from the list.
+
+  3. If there are no longer any `InternalIP` or `ExternalIP` elements
+     in the list, then the provided `--node-ips` is invalid, and
+     kubelet will error out.
+
+  4. If either element of `--node-ips` is a specific IP address, and
+     that IP address does not appear in the address list, then the
+     provided `--node-ips` is invalid and kubelet will error out.
+
+     Likewise, if the indicated IP could not possibly be chosen as the
+     Primary/Secondary Node IP because it is an `ExternalIP` and would
+     always be ignored in favor of an `InternalIP`, then the provided
+     `--node-ips` is invalid and kubelet will error out.
+
+  5. At this point kubelet will determine what the Primary Node IP and
+     (if relevant) Secondary Node IP would be based on the current
+     address list. If the results are compatible with what is
+     specified in `--node-ips`, then we are done.
+
+  6. Otherwise, kubelet will look for an element of the address list
+     that satisfies the first element of `--node-ips`, and (if
+     `--node-ips` has two elements) an element that satisfies the
+     second element of `--node-ips`, and move it/them to the front of
+     the list, in the correct order. It is guaranteed at this point
+     that at least one of those two elements exists, and that moving
+     it/them will result in a list of addresses that matches
+     `--node-ips`.
+
+  7. The resulting list is then assigned to `Node.Status.Addresses`.
+
+Eg, suppose `--node-ips` is `ipv4,ipv6` and the initial list of
+addresses is `[ fd01::1, fd01::2, 10.0.0.1, 10.0.0.2 ]`. In that case
+steps 1-4 would have no effect. Step 5 would determine that the
+Primary Node IP was `fd01::1` and the Secondary Node IP was
+`10.0.0.1`, which does not match `ipv4,ipv6`. (Had we said `ipv6,ipv4`
+instead we'd be done though.) Thus, step 6 would reorder the list to
+`[ 10.0.0.1, fd01::1, fd01::2, 10.0.0.2 ]`, and then the Primary Node
+IP would be `10.0.0.1` and the Secondary would be `fd01::1`, and we
+would match.
+
+With the same list of addresses but `--node-ips 10.0.0.2`, step 2
+would remove the two IPv6 addresses, step 4 would confirm that
+`10.0.0.2` was in the list, step 5 would find the wrong Primary Node
+IP, and step 6 would move the desired IP to the front of the list,
+giving `[ 10.0.0.2, 10.0.0.1 ]`.
+
+With `--node-ips ipv4,ipv6` and initial list of addresses `[ 10.0.0.1,
+10.0.0.2 ]`, steps 1-4 would have no effect, and step 5 would find
+that the Primary Node IP was `10.0.0.1` and there was no Secondary
+Node IP. This _is_ compatible with `--node-ips ipv4,ipv6` (since both
+elements are optional), so we would keep the list of node addresses
+unchanged. Note that we would get exactly the same result with
+`--node-ips ipv6,ipv4`.
+
+```
+<<[UNRESOLVED reordering ]>>
```

Comment:
I was trying to design this in a way so that kubelet would do the right thing automatically in most circumstances, but yes, if we forced the user to specify more, we could simplify the proposal.